text/0000-canonical-metadata.md

```diff
@@ -0,0 +1,1466 @@
+        * Public methods
+    * Imports
+        * Builtin modules
+        * External modules
+        * Salesforce scoped modules
+        * Other modules(components and libraries)
+        * Relative imports
+        * Dynamic imports
+    * Exports
+        * Default exports
+        * Named exports
+        * Aggregating modules
+    * JSDoc
+        * Class
+        * Properties
+        * Public methods
+    * Programmatically generated events
+        * type
+        * options: composed, bubbles
+    * Programmatically fetched static resources
+        * `<script>`
+        * `<img>`
+        * `<link>`
+* CSS file
+    * Tokens
+    * Static resources
+
+## Metadata shape
+
+### TypeScript
+
+#### Bundle metadata shape
+```ts
+interface BundleMetadata {
+    version: string;
+    name?: string;
+    namespace?: string;
+    // Define the type of module that is represented by the given metadata object.
+    moduleType?: "HTML" | "CSS" | "Javascript";
+
+    /**
+     * For a component bundle, this property is a reference to the main component class in the bundle.
+     * This class must be the default export of the javascript file. The file must be named the same
+     * as the bundle
+     */
+    defaultComponentClass?: Class;
+
+    // Information about HTML template files in the bundle.
+    templates?: HTMLTemplateFile[];
+
+    // Information about javascript files in the bundle.
+    scripts?: ScriptFile[];
+
+    // Information about css files in the bundle.
+    css?: CSSFile[]
+}
+```
+Other alternatives:
+- Indexed by file type and file name
+```ts
+interface BundleMetadataOption2 {
+    version: string;
+    name?: string;
+    namespace?: string;
+    moduleType?: "HTML" | "CSS" | "Javascript";
+
+    defaultComponentClass?: ComponentClass[];
+
+    templates?: { [key: string]: HTMLTemplateFile[]};
+
+    scripts?: { [key: string]: ScriptFile[]};
+
+    css?: { [key: string]: CSSFile[]};
+}
+```
+
+#### HTML Template file metadata shape
+```ts
+// Root metadata object for an HTML template file
+interface HTMLTemplateFile {
+    fileType: 'html';
+    fileName: string;
+    // List of unique components referenced in the template
+    componentReferences?: ComponentReference[];
+    // Static resources referenced in a template. For example: 'src' attribute value of an <img>, <source> tag.
+    staticResources?: StaticResourceReference[];
+    // Template in AST format starting with the root <template> node. Note: this is a partial AST
+    astRoot?: Node;
+}
+
+// Represents a component referenced in the template.
+// Use the 'tagName' property for kebab case and 'name' property for the reference in canonical form
+interface ComponentReference extends ModuleReference {
+    // Component reference always has a namespace
+    namespace: string;
+    // No moduleIdentifier for a component
+    moduleIdentifier?: never;
+    // Components are always imported from an external bundle
+    type: 'external';
+    tagName: string;
+}
```

Comment:
how does ComponentReference shape reflect the non-default namespace for which the current file is being compiled? As it stands, the metadata collection is done on a 'as authored' raw source and doesn't take custom namespace into account. For comparison, in the existing lwc-compiler meta collection flow, all the [Reference](https://git.soma.salesforce.com/lwc/lwc-api/blob/master/src/main/java/com/salesforce/lwc/reference/Reference.java#L20) objects surface 'namespacedId' property, which reflects the mapping from default to custom namespace. In addition, there is an assumption that all component dependencies, including template dependencies, will be namespaced by post-processing the compiler output, in 'amd' format, by applying the namespace mapping to the array of 'amd' module dependencies. 
Now back to the current proposal, the new metadata package will be invoked stand alone, outside of any namespace specific logic that resides in the compiler, thus the responsiblity for any namespace manipuations will have to be done by the consumer. With that, should we ease the namespacing pain and allow the meta package to accept an optional configuration to apply namespacing?

I propose we do. For example, if we encounter c-view-source template reference and the meta collection is invoked with a namespace mapping from the default 'c' to 'foo', then the namespace should be applied on all  ComponentReference properties. The question is, how do we surface it? Do we apply the namespace on the existing 'name' and 'tagName' properties or create the new ones? 

_template reference c-view-source_
_Example of an output with namespace applied to the existing props:_
```js
        {
            "name": "foo/viewSource",
            "namespace": "foo",
            "type": "external",
            "tagName": "foo-view-source"
        }
```
_Example of an output with adding namespacing as a new optional 'namespaceMapping' property_:
```js
        {
            "name": "c/viewSource",
            "namespace": "c",
            "type": "external",
            "tagName": "c-view-source",
            "namespaceMapping": {
                  "name": "foo/viewSource", 
                  "tagName": "foo-view-source"
             }
        }
```

Reply:
Per offline discussion, metadata will be namespace agnostic and will have to be performed by the consumed. Marking this conversation as resolved.